src/session.c

```diff
@@ -7,33 +7,168 @@
 /* Session stuff */
 struct p11prov_session {
     P11PROV_CTX *provctx;
+    P11PROV_SESSION_POOL *pool;
 
     CK_SLOT_ID slotid;
     CK_SESSION_HANDLE session;
 
     int refcnt;
+    int free;
 };
 
-P11PROV_SESSION *p11prov_session_new(P11PROV_CTX *ctx, CK_SLOT_ID slotid)
+struct p11prov_session_pool {
+    P11PROV_CTX *provctx;
+
+    CK_ULONG max_sessions;
+    CK_ULONG limit_sessions;
+    CK_ULONG cur_sessions;
+
+    P11PROV_SESSION **sessions;
+    int num_sessions;
+
+    P11PROV_SESSION *login_session;
+
+    pthread_mutex_t lock;
+};
+
+CK_RV p11prov_session_pool_init(P11PROV_CTX *ctx, CK_TOKEN_INFO *token,
+                                P11PROV_SESSION_POOL **_pool)
+{
+    P11PROV_SESSION_POOL *pool;
+
+    pool = OPENSSL_zalloc(sizeof(P11PROV_SESSION_POOL));
+    if (!pool) {
+        return CKR_HOST_MEMORY;
+    }
+    pool->provctx = ctx;
+
+    pthread_mutex_init(&pool->lock, 0);
+
+    if (token->ulMaxSessionCount != 0) {
+        pool->max_sessions = token->ulMaxSessionCount;
+        /* keep a max of 10% of the sessions */
+        pool->limit_sessions = pool->max_sessions / 10;
+        if (pool->limit_sessions == 0) {
+            pool->limit_sessions = 3;
+        }
+    } else {
+        /* arbitrary max concurrent open sessions */
+        pool->max_sessions = 1024;
+        pool->limit_sessions = 8;
+    }
+    if (pool->limit_sessions > pool->max_sessions) {
+        pool->limit_sessions = pool->max_sessions;
+    }
+
+    *_pool = pool;
+    return CKR_OK;
+}
+
+CK_RV p11prov_session_pool_free(P11PROV_SESSION_POOL *pool)
+{
+    P11PROV_SESSION *session;
+
+    if (!pool) {
+        return CKR_OK;
+    }
+
+    /* deref login_session first */
+    p11prov_session_free(pool->login_session);
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&pool->lock);
+    for (int i = 0; i < pool->num_sessions; i++) {
+        session = pool->sessions[i];
+        if (session->refcnt > 1) {
+            P11PROV_raise(pool->provctx, CKR_GENERAL_ERROR,
+                          "Can't free multiply-referenced session (%u)", i);
```

Comment:
This still frees the object. Either add `continue`, or maybe adjust error to `Freeing still referenced session` to reflect the action taken better?

Reply:
Will add a way to orphan the entry and move on